### ISSUE TITLE
Overfitting: validation loss stuck.

### ISSUE THREAD

**Issue by ksaidin** [NVIDIA/tacotron2#344]
Hello,
I am experimenting with 3.5 hours of data set. (16kbit, 22khz, 9.5s av.duration, 1300 audio files)
I am using pretrained models which are provided by this repo.

My hparams;
p_attention_dropout=0.5,
p_decoder_dropout=0.5,
learning_rate=4e-5,
batch_size=16,

After 15k iteraton: Train loss 0.25. It is producing almost original like outputs.
![alignment](https://user-images.githubusercontent.com/6515490/80947565-635d0580-8e2b-11ea-96b2-a9bd8d424a10.JPG)
![melpredict](https://user-images.githubusercontent.com/6515490/80947577-6a841380-8e2b-11ea-8619-a9a6fac67c68.png)
![meltarget](https://user-images.githubusercontent.com/6515490/80947589-71128b00-8e2b-11ea-954d-0721065ccb40.png)

The only thing I couldn't improve is a validation loss, it stuck since 4000 iteration.

![validationloss](https://user-images.githubusercontent.com/6515490/80948100-65739400-8e2c-11ea-8843-2d58cc1d9178.JPG)

All suggestions are welcome! Thanks beforehand.

**Comment by rafaelvalle** [comment]
Very nice attention map!
My guess is that you're model will start overfitting the training data, which implies you either need to augment your data or regularize the model more strongly, e.g. increase weight decay or decrease the prenet dimensions.

**Comment by ksaidin** [comment]
Thank you for your recommendations @rafaelvalle ! 

I've been tuning weight decay and prenet dimensions with no success, I will go and try inflating my data.

**Comment by ksaidin** [comment]
After I artificially boosted my training data, (up to 6,5 hrs) I've managed to decrease my validation loss to ~0.25. Now I have roughly equal training and validation losses.

My attention plot after 18.5K iterations as below. I am not sure if I am reading it correctly. What are those horizontal lines on the map?

![crcl](https://user-images.githubusercontent.com/6515490/81452476-d0b0c380-91a8-11ea-832b-b2126ce3305b.jpg)

**Comment by CookiePPP** [comment]
@ksaidin 
The horizontal lines show periods where the model will be generating off of the same piece of text.
Typically these horizontal lines will typically be arounds commas or periods where the model **should** be pausing.
You'll notice these lines are blurred, this would be for 2 reasons.
1. The model cannot tell the difference between spaces (" ") and commas (",") yet, so when two are neighbouring it will just send a mix of both.
2. It is hard to align along 200 timesteps where most of the inputs are either silence or breathing, very little information to help the model figure out where it is.

Once your model stops improving, you can drop your learning rate and see if the model is able to align along those long stretches.
To my memory, the model needs about 1e-4 learning rate to align along pauses (though ideal learning rate is batch size dependent)

Also be aware that small periods of low alignment can make the entire model unstable when you actually go to use it.
**During Validation**
![TeacherForcedAlignment](https://user-images.githubusercontent.com/42448678/81457518-51ae9180-918e-11ea-87aa-0b1b8405d416.png)
**Same text without any Teacher Forcing (like how you'd use the model for new text)**
![EffectivelyInferrenceAlignment](https://user-images.githubusercontent.com/42448678/81457521-5410eb80-918e-11ea-934a-3e9ee26ecb74.png)

**Comment by ksaidin** [comment]
Thank you @CookiePPP for explanation.

I was too worried about inner silence, because there are thousands of audio files and usually the silence periods inside are not same. I had significant improvements after @Yeongtae 's [trimming code](https://github.com/Yeongtae/tacotron2/blob/master/preprocess_audio.py). I was inspired by that work and I've worked out [my version](https://github.com/ksaidin/tacotron2/blob/master/preprocess_inside_silence.py) of removing inside silence. I've just started training my dataset after then, I don't how much it helps.

**Comment by CookiePPP** [comment]
@ksaidin 
The attention plot you showed has 2 seconds of time spent on the same text piece, so I think that you will find some improvement removing some of the inner silence.
Good Luck regardless!

**Comment by ksaidin** [comment]
Thank you @CookiePPP , Attention plot's aligned faster this time,

![4k](https://user-images.githubusercontent.com/6515490/81475694-976f6680-922f-11ea-9392-1549697d52c7.JPG)

Why do you think this anomaly in train loss. It jumped high up for some time despite nothing's changed in parameters. I've just let it run and it normalized later on.

![Uploading traininanomaly.JPG…]()

**Comment by rafaelvalle** [comment]
@ksaidin hope did you augment the data?

**Comment by ksaidin** [comment]
@rafaelvalle , in two ways: I've created new audios by parsing from different starting points than older ones and I've added random noise.

**Comment by ksaidin** [comment]
Status report, after 27k iteration I have ~0.22 training loss and ~0.21 validation loss. I have annealed my learning rate to 5e-5. Weight decay stays 1e-6 since I have my validation loss is decreasing alongside with training loss equally.

My attention map has disconnections though output is stable, I guess I will keep decreasing my learning rate.
![27000](https://user-images.githubusercontent.com/6515490/81604290-7eaeae80-93f1-11ea-9293-02d96fa08e85.JPG)

**Comment by ksaidin** [comment]
Another anomaly, grad norm jump as high as 1270! somewhere around 33k iteration. This experiment is showing me really weird graphs. Anyone can explain?

![gradnormanomaly](https://user-images.githubusercontent.com/6515490/81732631-9306b000-94b2-11ea-9df0-e91f3be672ef.jpg)

**Comment by ksaidin** [comment]
@rafaelvalle , is there a minimum limit for learning rate? My learning rate approaches 1e-6 and it seems my grad.norm calculation is broken.

I was planning to run this experiment for 20K to 30K more iterations on 1e-6. Now I've stopped to understand what's going wrong.

**Comment by rafaelvalle** [comment]
Around 1e-6 you might start getting numerical precision issues that can make your loss explode.

**Comment by ksaidin** [comment]
Thank you @rafaelvalle !

Conclusions on training small dataset:

I've got better results after data augmentation than penalizing model complexity. Weight regularization definitely helps reduce overfitting but data augmentation evens losses.

I kept annealing my learning rate manually; reduced 10% when I see no regress for ~2000 iterations, It worked for me till I reach 1e-5.

My final graphs;

![trainloss_experiemnt](https://user-images.githubusercontent.com/6515490/81920112-8db47d00-95fa-11ea-83b3-a0a521275b17.JPG)

![validationloss_experiemtn](https://user-images.githubusercontent.com/6515490/81920125-91e09a80-95fa-11ea-943b-d0009bb72b85.JPG)

![alignment_experiment](https://user-images.githubusercontent.com/6515490/81920137-94db8b00-95fa-11ea-9503-64214ce82ade.JPG)

Thank you.

**Comment by EuphoriaCelestial** [comment]
> @rafaelvalle , in two ways: I've created new audios by parsing from different starting points than older ones and I've added random noise.

If I augment the data by adding noise, will it affect the output audio quality?

**Comment by rafaelvalle** [comment]
Make sure your noise is zero mean and equally distributed around the mean.

**Comment by shaktikshri** [comment]
@ksaidin , I might be late to ask this but what was your execution time for 15k steps? I have a very similar dataset as yours (5.4hrs). My stats are at 14sec/step on V100 GPU, batch size 128. but i've added some latent variables which are part of the model. Trying to figure out if i have an extremely slow implementation or is it just normal?

**Comment by EuphoriaCelestial** [comment]
> @ksaidin , I might be late to ask this but what was your execution time for 15k steps? I have a very similar dataset as yours (5.4hrs). My stats are at 14sec/step on V100 GPU, batch size 128. but i've added some latent variables which are part of the model. Trying to figure out if i have an extremely slow implementation or is it just normal?

that was slow, I trained my model on RTX 2080ti, batch_size=32, each step take around 1,5s

**Comment by ksaidin** [comment]
@shaktikshri I don't understand what's your worry exactly about? If your training data average length is 2 seconds then you will have a smaller time per step than having training data average length of 18 seconds. Am I missing something?

**Comment by EuphoriaCelestial** [comment]
@ksaidin can you please share your data augment code?

**Comment by pathnirvana** [comment]
I am training with a small Sinhala (indic language of Sri Lanka) language dataset of 1.5 hours and observe overfitting as expected. (I am in the process of creating a bigger dataset but it will take few weeks) 

However unlike @ksaidin 's results above I do not get good alignment. Also when I synthesis speech with a some text in the training set with 500 epoch checkpoint I hear garbage - my thinking was that I should here something intelegible since I was using a prompt from the training set. Please see the results below and let me know what I might be doing wrong. 

**important details**

1. I use the same audio format as LJ dataset and all the audio hparams are the same
2. however I modify the symbols to `symbols = [_pad] + list(_special) + list(_punctuation) + list('abcdefghijklmnopqrstuvwxyzDT')` and transliterate sinhala characters based on the pronounciation to match these symbol range. - no cleaners used
3. for synthesis I am using pretrained waveglow v5 model 

![tts help needed](https://user-images.githubusercontent.com/18088975/105958045-ac857180-60ac-11eb-8988-d067eec32f3a.png)

**Comment by ramin40** [comment]
> I am training with a small Sinhala (indic language of Sri Lanka) language dataset of 1.5 hours and observe overfitting as expected. (I am in the process of creating a bigger dataset but it will take few weeks)
> 
> However unlike @ksaidin 's results above I do not get good alignment. Also when I synthesis speech with a some text in the training set with 500 epoch checkpoint I hear garbage - my thinking was that I should here something intelegible since I was using a prompt from the training set. Please see the results below and let me know what I might be doing wrong.
> 
> **important details**
> 
>     1. I use the same audio format as LJ dataset and all the audio hparams are the same
> 
>     2. however I modify the symbols to `symbols = [_pad] + list(_special) + list(_punctuation) + list('abcdefghijklmnopqrstuvwxyzDT')` and transliterate sinhala characters based on the pronounciation to match these symbol range. - no cleaners used
> 
>     3. for synthesis I am using pretrained waveglow v5 model
> 
> 
> ![tts help needed](https://user-images.githubusercontent.com/18088975/105958045-ac857180-60ac-11eb-8988-d067eec32f3a.png)



> I am training with a small Sinhala (indic language of Sri Lanka) language dataset of 1.5 hours and observe overfitting as expected. (I am in the process of creating a bigger dataset but it will take few weeks)
> 
> However unlike @ksaidin 's results above I do not get good alignment. Also when I synthesis speech with a some text in the training set with 500 epoch checkpoint I hear garbage - my thinking was that I should here something intelegible since I was using a prompt from the training set. Please see the results below and let me know what I might be doing wrong.
> 
> **important details**
> 
>     1. I use the same audio format as LJ dataset and all the audio hparams are the same
> 
>     2. however I modify the symbols to `symbols = [_pad] + list(_special) + list(_punctuation) + list('abcdefghijklmnopqrstuvwxyzDT')` and transliterate sinhala characters based on the pronounciation to match these symbol range. - no cleaners used
> 
>     3. for synthesis I am using pretrained waveglow v5 model
> 
> 
> ![tts help needed](https://user-images.githubusercontent.com/18088975/105958045-ac857180-60ac-11eb-8988-d067eec32f3a.png)

Hi there. I have the same issue. have you found any reason or solution?
